File: services/backend/pkg/models/orgs.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/uptrace/bun"
)

type StringList []string

func (s StringList) Value() (driver.Value, error) {
	if s == nil {
		return "[]", nil
	}
	b, err := json.Marshal(s)
	return string(b), err
}

func (s *StringList) Scan(v interface{}) error {
	var b []byte
	switch t := v.(type) {
	case []byte:
		b = t
	case string:
		b = []byte(t)
	default:
		return fmt.Errorf("unexpected type %T", v)
	}
	return json.Unmarshal(b, s)
}

type Org struct {
	bun.BaseModel `bun:"table:orgs"`

	ID              uuid.UUID  `bun:",pk,type:uuid,default:gen_random_uuid()" json:"id"`
	Name            string     `bun:"name,type:text,notnull,unique" json:"name"`
	Description     string     `bun:"description,type:text,default:''" json:"description"`
	ImagePatterns   StringList `bun:"image_patterns,type:jsonb,default:'[]'" json:"image_patterns"`
	CreatedByID     uuid.UUID  `bun:"created_by_id,type:uuid,notnull" json:"created_by_id"`
	CreatedAt       time.Time  `bun:"created_at,type:timestamptz,default:now()" json:"created_at"`
	UpdatedAt       time.Time  `bun:"updated_at,type:timestamptz" json:"updated_at"`
	CurrentUserRole string     `bun:"-" json:"current_user_role,omitempty"`

	Policies []OrgPolicy `bun:"rel:has-many,join:id=org_id" json:"policies,omitempty"`
}

const (
	OrgRoleOwner  = "owner"
	OrgRoleAdmin  = "admin"
	OrgRoleEditor = "editor"
	OrgRoleViewer = "viewer"
)

type OrgMember struct {
	bun.BaseModel `bun:"table:org_members"`

	OrgID     uuid.UUID `bun:"org_id,pk,type:uuid,notnull" json:"org_id"`
	UserID    uuid.UUID `bun:"user_id,pk,type:uuid,notnull" json:"user_id"`
	Role      string    `bun:"role,type:text,notnull,default:'viewer'" json:"role"`
	JoinedAt  time.Time `bun:"joined_at,type:timestamptz,notnull,default:now()" json:"joined_at"`
	CreatedAt time.Time `bun:"created_at,type:timestamptz,default:now()" json:"created_at"`
	UpdatedAt time.Time `bun:"updated_at,type:timestamptz,default:now()" json:"updated_at"`
	Email     string    `bun:"-" json:"email,omitempty"`
	Username  string    `bun:"-" json:"username,omitempty"`
}

type OrgInvite struct {
	bun.BaseModel `bun:"table:org_invites"`

	ID               uuid.UUID  `bun:",pk,type:uuid,default:gen_random_uuid()" json:"id"`
	OrgID            uuid.UUID  `bun:"org_id,type:uuid,notnull" json:"org_id"`
	Email            string     `bun:"email,type:text,notnull" json:"email"`
	Role             string     `bun:"role,type:text,notnull,default:'viewer'" json:"role"`
	Token            string     `bun:"token,type:varchar(64),notnull,unique" json:"token"`
	InvitedByUserID  uuid.UUID  `bun:"invited_by_user_id,type:uuid,notnull" json:"invited_by_user_id"`
	AcceptedByUserID *uuid.UUID `bun:"accepted_by_user_id,type:uuid" json:"accepted_by_user_id,omitempty"`
	AcceptedAt       *time.Time `bun:"accepted_at,type:timestamptz" json:"accepted_at,omitempty"`
	RevokedAt        *time.Time `bun:"revoked_at,type:timestamptz" json:"revoked_at,omitempty"`
	ExpiresAt        time.Time  `bun:"expires_at,type:timestamptz,notnull" json:"expires_at"`
	CreatedAt        time.Time  `bun:"created_at,type:timestamptz,default:now()" json:"created_at"`
	UpdatedAt        time.Time  `bun:"updated_at,type:timestamptz,default:now()" json:"updated_at"`
	OrgName          string     `bun:"-" json:"org_name,omitempty"`
	OrgDescription   string     `bun:"-" json:"org_description,omitempty"`
	InvitedByEmail   string     `bun:"-" json:"invited_by_email,omitempty"`
	InvitedByName    string     `bun:"-" json:"invited_by_username,omitempty"`
}

type ComplianceHistory struct {
	bun.BaseModel `bun:"table:compliance_history"`

	ID          uuid.UUID `bun:",pk,type:uuid,default:gen_random_uuid()" json:"id"`
	ScanID      uuid.UUID `bun:"scan_id,type:uuid,notnull" json:"scan_id"`
	PolicyID    uuid.UUID `bun:"policy_id,type:uuid,notnull" json:"policy_id"`
	OrgID       uuid.UUID `bun:"org_id,type:uuid,notnull" json:"org_id"`
	Status      string    `bun:"status,type:text,notnull" json:"status"`
	EvaluatedAt time.Time `bun:"evaluated_at,type:timestamptz,default:now()" json:"evaluated_at"`
}

type PolicyRule struct {
	Type     string  `json:"type"`               // max_cvss, max_count, max_total, require_fix, blocked_cve
	Value    float64 `json:"value,omitempty"`    // numeric threshold
	Severity string  `json:"severity,omitempty"` // for max_count, require_fix
	CVEID    string  `json:"cve_id,omitempty"`   // for blocked_cve
}

type PolicyRuleList []PolicyRule

func (r PolicyRuleList) Value() (driver.Value, error) {
	b, err := json.Marshal(r)
	return string(b), err
}

func (r *PolicyRuleList) Scan(v interface{}) error {
	var b []byte
	switch t := v.(type) {
	case []byte:
		b = t
	case string:
		b = []byte(t)
	default:
		return fmt.Errorf("unexpected type %T", v)
	}
	return json.Unmarshal(b, r)
}

type OrgPolicy struct {
	bun.BaseModel `bun:"table:org_policies"`

	ID        uuid.UUID      `bun:",pk,type:uuid,default:gen_random_uuid()" json:"id"`
	OrgID     uuid.UUID      `bun:"org_id,type:uuid,notnull" json:"org_id"`
	Name      string         `bun:"name,type:text,notnull" json:"name"`
	Rules     PolicyRuleList `bun:"rules,type:jsonb" json:"rules"`
	CreatedAt time.Time      `bun:"created_at,type:timestamptz,default:now()" json:"created_at"`
	UpdatedAt time.Time      `bun:"updated_at,type:timestamptz" json:"updated_at"`
}

type OrgScan struct {
	bun.BaseModel `bun:"table:org_scans"`

	OrgID  uuid.UUID `bun:"org_id,type:uuid,notnull" json:"org_id"`
	ScanID uuid.UUID `bun:"scan_id,type:uuid,notnull" json:"scan_id"`
}

type OrgRegistry struct {
	bun.BaseModel `bun:"table:org_registries"`

	OrgID      uuid.UUID `bun:"org_id,type:uuid,notnull" json:"org_id"`
	RegistryID uuid.UUID `bun:"registry_id,type:uuid,notnull" json:"registry_id"`
}

type OrgWatchlistItem struct {
	bun.BaseModel `bun:"table:org_watchlist_items"`

	OrgID           uuid.UUID `bun:"org_id,type:uuid,notnull" json:"org_id"`
	WatchlistItemID uuid.UUID `bun:"watchlist_item_id,type:uuid,notnull" json:"watchlist_item_id"`
}

type OrgTag struct {
	bun.BaseModel `bun:"table:org_tags"`

	OrgID uuid.UUID `bun:"org_id,type:uuid,notnull" json:"org_id"`
	TagID uuid.UUID `bun:"tag_id,type:uuid,notnull" json:"tag_id"`
}

type OrgSuppression struct {
	bun.BaseModel `bun:"table:org_suppressions"`

	OrgID         uuid.UUID `bun:"org_id,type:uuid,notnull" json:"org_id"`
	SuppressionID uuid.UUID `bun:"suppression_id,type:uuid,notnull" json:"suppression_id"`
}

type OrgStatusPage struct {
	bun.BaseModel `bun:"table:org_status_pages"`

	OrgID        uuid.UUID `bun:"org_id,type:uuid,notnull" json:"org_id"`
	StatusPageID uuid.UUID `bun:"status_page_id,type:uuid,notnull" json:"status_page_id"`
}

type Violation struct {
	Rule    PolicyRule `json:"rule"`
	Message string     `json:"message"`
	VulnID  string     `json:"vuln_id,omitempty"`
}

type ViolationList []Violation

func (v ViolationList) Value() (driver.Value, error) {
	b, err := json.Marshal(v)
	return string(b), err
}

func (v *ViolationList) Scan(src interface{}) error {
	var b []byte
	switch t := src.(type) {
	case []byte:
		b = t
	case string:
		b = []byte(t)
	default:
		return fmt.Errorf("unexpected type %T", src)
	}
	return json.Unmarshal(b, v)
}

type ComplianceResult struct {
	bun.BaseModel `bun:"table:compliance_results"`

	ID          uuid.UUID     `bun:",pk,type:uuid,default:gen_random_uuid()" json:"id"`
	ScanID      uuid.UUID     `bun:"scan_id,type:uuid,notnull" json:"scan_id"`
	PolicyID    uuid.UUID     `bun:"policy_id,type:uuid,notnull" json:"policy_id"`
	OrgID       uuid.UUID     `bun:"org_id,type:uuid,notnull" json:"org_id"`
	Status      string        `bun:"status,type:text,notnull" json:"status"` // pass, fail
	Violations  ViolationList `bun:"violations,type:jsonb" json:"violations"`
	EvaluatedAt time.Time     `bun:"evaluated_at,type:timestamptz,default:now()" json:"evaluated_at"`

	PolicyName string `bun:"-" json:"policy_name,omitempty"`
	OrgName    string `bun:"-" json:"org_name,omitempty"`
}

File: services/frontend/lib/api/types/orgs.ts
export interface PolicyRule {
  type:
    | 'max_cvss'
    | 'max_count'
    | 'max_total'
    | 'require_fix'
    | 'blocked_cve'
    | 'xray_policy_block';
  value?: number;
  severity?: string;
  cve_id?: string;
}

export interface OrgPolicy {
  id: string;
  org_id: string;
  name: string;
  rules: PolicyRule[];
  include_suppressed: boolean;
  created_at: string;
  updated_at: string;
}

export type OrgRole = 'owner' | 'admin' | 'editor' | 'viewer';

export type VulnerabilityViewSortBy =
  | 'vuln_id'
  | 'pkg_name'
  | 'severity'
  | 'cvss_score'
  | 'installed_version'
  | 'fixed_version';
export type VulnerabilityViewSortDir = 'asc' | 'desc';
export type VulnerabilityViewSeverity = '' | 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW' | 'UNKNOWN';

export interface VulnerabilityViewSettings {
  sort_by: VulnerabilityViewSortBy;
  sort_dir: VulnerabilityViewSortDir;
  severity: VulnerabilityViewSeverity;
  min_cvss: number;
  has_fix: boolean;
  xray_policy_first: boolean;
  policy_failed_only: boolean;
}

export interface VulnerabilityViewPreferenceResponse {
  settings: VulnerabilityViewSettings;
  source: 'system' | 'org' | 'user';
  scope_type: 'personal' | 'org';
  scope_ref: string;
  has_user_override: boolean;
  org_settings?: VulnerabilityViewSettings;
}

export interface Org {
  id: string;
  name: string;
  description: string;
  member_count?: number;
  policy_count?: number;
  scan_count?: number;
  unique_image_count?: number;
  last_scan_at?: string | null;
  is_active: boolean;
  allow_image_scans: boolean;
  allow_helm_scans: boolean;
  allow_rescans: boolean;
  allow_member_invites: boolean;
  allow_org_tokens: boolean;
  created_by_id: string;
  created_at: string;
  updated_at: string;
  current_user_role?: OrgRole;
  vulnerability_view_settings?: VulnerabilityViewSettings;
  policies?: OrgPolicy[];
}

export interface ResourceShare {
  org_id: string;
  org_name: string;
  org_description?: string;
  is_owner: boolean;
}

export interface OrgMember {
  org_id: string;
  user_id: string;
  role: OrgRole;
  joined_at: string;
  created_at: string;
  email?: string;
  username?: string;
}

export interface OrgInvite {
  id: string;
  org_id: string;
  org_name?: string;
  org_description?: string;
  email: string;
  role: Extract<OrgRole, 'admin' | 'editor' | 'viewer'>;
  token: string;
  invited_by_user_id: string;
  invited_by_email?: string;
  invited_by_username?: string;
  accepted_by_user_id?: string | null;
  accepted_at?: string | null;
  revoked_at?: string | null;
  expires_at: string;
  created_at: string;
  updated_at: string;
}

export interface TrendPoint {
  date: string;
  pass: number;
  fail: number;
  rate: number;
}

export interface Violation {
  rule: PolicyRule;
  message: string;
  vuln_id?: string;
}

export interface ComplianceResult {
  id: string;
  scan_id: string;
  policy_id: string;
  org_id: string;
  status: 'pass' | 'fail';
  violations: Violation[];
  evaluated_at: string;
  policy_name?: string;
  org_name?: string;
  policy_rules?: PolicyRule[];
}

export interface APIToken {
  id: string;
  key: string;
  description: string;
  type: string;
  disabled: boolean;
  disabled_reason: string;
  created_at: string;
  expires_at: string;
  user_id: string;
  org_id?: string;
  scope: OrgTokenScope;
}

export type OrgTokenScope = 'org_admin' | 'pipeline_scan';

export interface PersonalToken {
  id: string;
  description: string;
  type: string;
  disabled: boolean;
  disabled_reason: string;
  created_at: string;
  expires_at: string;
}

export interface AuditEntry {
  id: string;
  org_id: string;
  user_id: string;
  username: string;
  email: string;
  action: string;
  details: string;
  created_at: string;
}

export interface OrgRiskScore {
  score: number;
  grade: string;
  unique_images: number;
  total_scans: number;
  totals: {
    critical: number;
    high: number;
    medium: number;
    low: number;
    unknown: number;
  };
  compliance_pass_rate: number;
  compliance_pass: number;
  compliance_fail: number;
}

export interface PipelineScanHistoryItem {
  id: string;
  scan_id: string;
  source: string;
  external_ref?: string;
  delivery_status: string;
  delivery_attempt_count: number;
  last_delivery_error?: string;
  last_attempt_at?: string | null;
  delivered_at?: string | null;
  created_at: string;
  initiator?: {
    source: string;
    token_id?: string | null;
    token_description?: string;
  };
  scan: {
    id: string;
    image_name: string;
    image_tag: string;
    status: string;
    current_step: string;
    critical_count: number;
    high_count: number;
    completed_at?: string | null;
  };
}
